storage/durability/durability.rs

```diff
@@ -15,62 +15,56 @@
+use std::{
+    error::Error,
+    fmt::{Display, Formatter},
+    io::{self, Read, Write},
+};
 
 use primitive::u80::U80;
+use serde::{Deserialize, Serialize};
 
 pub mod wal;
 
-///
-/// Notes:
-///     We can provide to a DurabilityService for Records
-///     Records have a serialised Byte representations and type (ID)
-///
-///     We have two choices:
-///         1. Long files (~ 100s MB) with many records embedded into fixed-sized blocks (fixed size = seekable)
-///         2. Shorter files (~ 10s MB) without fixed sized blocks internally.
-///
-///     We should aim to minimise the amount of data ending on the Disk, which means compression is critical.
-///     For this, having rolling shorter files that are compressed in larger, irregular chunks as they are written,
-///     is more optimal. Having rolling shorter files named by the first sequence number they contain,
-///     still allows skipping through the logical WAL by binary searching file names,
-///     then doing a scan through 10s of MB of data. This is acceptable if:
-///         1. we rarely need to do point look ups in data (iterating is amortises the cost)
-///         2. the costs of reading 10s of extra MB during a lookup is acceptable to overall performance.
-///
-///     Either solution will allow writing SerialisedRecord (data + type tag), and retrieving SerialisedRecords.
-///
-///     The final piece of the puzzle will be: how do we ensure with very high likelihood that we don't have two
-///     different implementations of SerialisedRecord using the same type tag?
-///
-///     First idea:
-///         On creation, we register a "name" and "type ID" against the DurabilityService.
-///         The service only serialises Records of a TypeID + Name that have been recognised - this can fail with a debug_assert
-///         The service should check that all TypeIDs registered have a different Name.
-///
-/// Other questions:
-///     1. Recovery/Checksum requirements - what are the failure modes
-///     2. How to benchmark
+pub type Result<T> = std::result::Result<T, DurabilityError>;
+
+#[derive(Debug)]
+struct RecordHeader {
+    sequence_number: SequenceNumber,
+    len: u32,
+    record_type: DurabilityRecordType,
+}
+
+#[derive(Debug)]
+pub struct RawRecord {
+    pub sequence_number: SequenceNumber,
+    pub record_type: DurabilityRecordType,
+    pub bytes: Box<[u8]>,
+}
 
 pub trait DurabilityService: Sequencer {
-    fn register_record_type(&mut self, record_type: DurabilityRecordType, record_name: &'static str);
+    fn register_record_type<Record: DurabilityRecord>(&mut self);
+
+    fn sequenced_write<Record>(&self, record: &Record) -> Result<SequenceNumber>
+    where
+        Record: DurabilityRecord;
 
-    fn sequenced_write(&self, record: &impl DurabilityRecord, record_name: &'static str) -> Result<SequenceNumber, DurabilityError>;
+    fn iter_from(&self, sequence_number: SequenceNumber) -> io::Result<impl Iterator<Item = io::Result<RawRecord>>>;
```

Comment:
let's add `iter_type_from` which takes a `DurabilityRecord` type and only returns deserialised records that have the right type, it shoudl be an iterator of `Result<(SequenceNumber, DurabilityRecord)>`
, where result is only for actual errors.